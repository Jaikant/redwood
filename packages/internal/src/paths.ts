import path from 'path'

import {
  getConfig,
  getConfigPath,
  getConfigSides,
  getSideConfig,
} from './config'
import { TargetEnum } from './config'

export interface NodeTargetPaths {
  db: string
  dbSchema: string
  src: string
  functions: string
  graphql: string
  lib: string
  services: string
}

export interface BrowserTargetPaths {
  src: string
  routes: string
  pages: string
  components: string
  layouts: string
  config: string
}

export interface Paths {
  base: string
  sides: {
    [side: string]: NodeTargetPaths | BrowserTargetPaths
  }
}

const mapNodePaths = (wsPath: string): NodeTargetPaths => {
  return {
    src: path.join(wsPath, 'src'),
    functions: path.join(wsPath, 'src/functions'),
    graphql: path.join(wsPath, 'src/graphql'),
    lib: path.join(wsPath, 'src/lib'),
    services: path.join(wsPath, 'src/services'),
    db: path.join(wsPath, 'prisma'),
    dbSchema: path.join(wsPath, 'prisma/schema.prisma'),
  }
}

const mapBrowserPaths = (wsPath: string): BrowserTargetPaths => {
  return {
    src: path.join(wsPath, 'src'),
    routes: path.join(wsPath, 'src/Routes.js'),
    pages: path.join(wsPath, 'src/pages'),
    components: path.join(wsPath, 'src/components'),
    layouts: path.join(wsPath, 'src/layouts'),
    config: path.join(wsPath, 'src/config'),
  }
}

/**
 * Absolute paths for the directory structure of a Redwood project based
 * on the `redwood.toml` file.
 */
export const getPaths = (): Paths => {
  // The Redwood config file denotes the base directory of a Redwood project.
  const base = path.dirname(getConfigPath())

  const configSides = getConfigSides()
  // Redwood supports different targets for sides. They have different directory
  // structures, so we map the side based on the "target" parameter.
  const sides = Object.keys(configSides).reduce((acc, key) => {
    const side = configSides[key]
    let paths
    switch (side.target) {
      case TargetEnum.NODE:
        paths = mapNodePaths(path.join(base, side.path))
        break
      case TargetEnum.BROWSER:
        paths = mapBrowserPaths(path.join(base, side.path))
        break
      default:
        throw new Error(
          `Woah there! "${key}" has a target that is is not currently supported:\n${JSON.stringify(
            side,
            undefined,
            2
          )}`
        )
    }
    return {
      [key]: paths,
      ...acc,
    }
  }, {})

  return {
    base,
    sides,
  }
}

export const getSidePaths = (
  name: string
): NodeTargetPaths | BrowserTargetPaths => {
  const paths = getPaths()
  if (!paths.sides[name]) {
    throw new Error(
      `A side named "${name}" does not exist? Is it in your redwood.toml configuration?`
    )
  }
  return paths.sides[name]
}
