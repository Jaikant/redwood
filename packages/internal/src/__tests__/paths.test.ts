import MockProject from '@redwoodjs/test-mocks'

import { getPaths, getSidePaths } from '../paths'

describe('paths', () => {
  const mockrw = new MockProject()

  beforeEach(() => {
    mockrw.mock()
  })

  afterAll(() => {
    mockrw.restore()
  })

  describe('getPaths', () => {
    it('provides defaults based off an empty configuration', () => {
      const paths = getPaths()
      mockrw.restore()
      expect(paths).toMatchSnapshot()
    })
  })

  describe('getSidePaths', () => {
    it('returns the correct paths for the given side', () => {
      const web = getSidePaths('web')
      mockrw.restore()
      expect(web).toMatchSnapshot()
    })

    it('throws when the side is incorrect', () => {
      expect(() => getSidePaths('banana')).toThrow()
    })
  })
})
