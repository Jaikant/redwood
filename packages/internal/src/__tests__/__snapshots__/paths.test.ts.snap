// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`paths getPaths provides defaults based off an empty configuration 1`] = `
Object {
  "base": "/mock/project",
  "sides": Object {
    "api": Object {
      "base": "/mock/project/api",
      "db": "/mock/project/api/prisma",
      "dbSchema": "/mock/project/api/prisma/schema.prisma",
      "functions": "/mock/project/api/src/functions",
      "graphql": "/mock/project/api/src/graphql",
      "lib": "/mock/project/api/src/lib",
      "services": "/mock/project/api/src/services",
      "src": "/mock/project/api/src",
    },
    "web": Object {
      "base": "/mock/project/web",
      "components": "/mock/project/web/src/components",
      "config": "/mock/project/web/src/config",
      "layouts": "/mock/project/web/src/layouts",
      "pages": "/mock/project/web/src/pages",
      "routes": "/mock/project/web/src/Routes.js",
      "src": "/mock/project/web/src",
    },
  },
}
`;

exports[`paths getSidePaths returns the correct paths for the given side 1`] = `
Object {
  "base": "/mock/project/web",
  "components": "/mock/project/web/src/components",
  "config": "/mock/project/web/src/config",
  "layouts": "/mock/project/web/src/layouts",
  "pages": "/mock/project/web/src/pages",
  "routes": "/mock/project/web/src/Routes.js",
  "src": "/mock/project/web/src",
}
`;
